bee-ternary/src/convert.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2020 IOTA Stiftung
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+// the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and limitations under the License.
+
+use crate::{Btrit, RawEncoding, RawEncodingBuf, TritBuf, Trits};
+use num_traits::{AsPrimitive, Bounded, FromPrimitive, One, ToPrimitive, Zero};
+use std::{
+    convert::TryFrom,
+    ops::{Add, Mul, Neg},
+};
+
+#[derive(Debug, PartialEq)]
+pub enum Error {
+    Empty,
+    TooBig,
+}
+
+impl<I: AsPrimitive<i64>, T: RawEncodingBuf> From<I> for TritBuf<T>
+where
+    T::Slice: RawEncoding<Trit = Btrit>,
+{
+    fn from(x: I) -> Self {
+        integer_trits(x).collect()
+    }
+}
+
+macro_rules! try_from_trits {
+    ($int:ident) => {
+        impl<'a, T: RawEncoding<Trit = Btrit> + ?Sized> TryFrom<&'a Trits<T>> for $int {
+            type Error = Error;
+            fn try_from(trits: &'a Trits<T>) -> Result<Self, Self::Error> {
+                trits_to_int(trits)
+            }
+        }
+    };
+}
+
+try_from_trits!(i64);
+try_from_trits!(i32);
+try_from_trits!(i16);
+try_from_trits!(i8);
+
+pub(crate) fn max_trits<I: ToPrimitive + Bounded>() -> usize {
+    (0.63 * std::mem::size_of::<I>() as f32 * 8.0) as usize + 1 // (log2/log3)*64
+}
+
+fn trits_to_int<I, T: RawEncoding<Trit = Btrit> + ?Sized>(trits: &Trits<T>) -> Result<I, Error>
+where
+    I: Copy + FromPrimitive + ToPrimitive + Bounded + Zero + One + Neg<Output = I> + Add + Mul,
+{
+    if trits.is_empty() {
+        Err(Error::Empty)
+    } else {
+        let max_trits: usize = max_trits::<I>();
+
+        assert!(max_trits == 41, "{}", max_trits);
```

Comment:
I'm not sure to understand this assert

Reply:
Whoops, that should have been gotten rid of. It's there as a test.